Control Center: Add missing Options, and Fix some

iamanaws opened the pull request.

iamanaws committed:
controlcenter: use string enums for consistency; add Spotlight module

- Replace boolean options with descriptive string enums ("show"/"hide",
  "whenActive"/"always"/"hide", "both"/"menuBar"/"controlCenter"/"hide")
- Add AccessibilityShortcuts, Battery, BatteryShowEnergyMode, Hearing,
  KeyboardBrightness, MusicRecognition, ScreenMirroring, StageManager,
  UserSwitcher, WiFi options
- Deprecate boolean values with backward-compatible warnings
- Add system.defaults.Spotlight.MenuItemHidden option

Iamanaws commented:
I just have doubt about the options that change their type, I know there are ways, but here I'm not sure how to add a warning or heads up, to indicate the allowed values had changed

bew commented on modules/system/defaults/controlcenter.nix:
```diff
@@ -1,100 +1,397 @@
+      '';
+    };
 
-            Show a Focus control in menu bar. Default is null.
```
(for all options)
I think we should keep the `Show a Focus control in menu bar` line in the description (or something similar), just having the path to the option in settings doesn't give much information..
It's still not very informative though since it's not clear what a 'focus control' is, but it's better than nothing 🤔

Iamanaws replied:
I added descriptions for all options to make it more clear. Also, I leave the Setting paths, and try to make it more clear what it is mapping to.

```nix
description = ''
        Show an Accessibility Shortcuts control in menu bar.

        This option mirrors the setting found in:
          System Preferences > Control Center > AccessibilityShortcuts

        Available settings:
          both           - Show in both the Menu Bar and Control Center (3)
          menuBar        - Show in the Menu Bar only (6)
          controlCenter  - Show in Control Center only (9)
          hide           - Don't Show (12)
'';
```
 
 
 We could move this to the bottom of the description to make it aside, but they were there before, and I think they still have value to know what are you doing.

```nix
description = ''
        Show an Accessibility Shortcuts control in menu bar.
        
        Available settings:
          both           - Show in both the Menu Bar and Control Center (3)
          menuBar        - Show in the Menu Bar only (6)
          controlCenter  - Show in Control Center only (9)
          hide           - Don't Show (12)
        
       This option mirrors the setting found in:
          System Preferences > Control Center > AccessibilityShortcuts
 '';
```

Iamanaws replied:
@bew 

Great, it's done

bew commented on modules/system/defaults/controlcenter.nix:
```diff
@@ -1,100 +1,397 @@
+
+        Show a bluetooth control in menu bar. Default is null.
+
+        18 = Show in Menu Bar
+        24 = Don't Show in Menu Bar
```
```suggestion
        true = Show in Menu Bar (18)
        false = Don't Show in Menu Bar (24)
```
Although, for a simple bool I'm not sure it's very useful to describe the option like this 👀 The description is enough for me.

bew commented on modules/system/defaults/controlcenter.nix:
```diff
@@ -1,100 +1,397 @@
+        Show a battery energy mode. Default is null.
+
+        false - "When Active"
+        true  - "Always"
```
Maybe don't use a boolean for this? an enum `whenActive` / `always` would be clearer.
OR rename the option like `...AlwaysShowBatteryEnergyMode` so the boolean makes sense 🤔

bew commented on modules/system/defaults/controlcenter.nix:
```diff
@@ -1,100 +1,397 @@
-            18 = Display icon in menu bar
-            24 = Hide icon in menu bar
-        '';
+        18 = Show in Menu Bar
+        24 = Don't Show in Menu Bar
```
```suggestion
        true = Show in Menu Bar
        false = Don't Show in Menu Bar
```
Or remove this section of the description?

bew commented on modules/system/defaults/controlcenter.nix:
```diff
@@ -1,100 +1,397 @@
+
+        Show a Wi-Fi control in menu bar. Default is null.
+
+        18 = Show in Menu Bar
+        24 = Don't Show in Menu Bar
```
```suggestion
        true = Show in Menu Bar (18)
        false = Don't Show in Menu Bar (24)
```

bew commented on modules/system/defaults/Spotlight.nix:
```diff
@@ -0,0 +1,18 @@
+
+        Show Spotlight in menu bar. Default is null.
+
+        false - "Don't Show in Menu Bar"
+        true  - "Show in Menu Bar"
```
```suggestion
        false - Don't Show in Menu Bar
        true  - Show in Menu Bar
```

bew reviewed (commented):
A few comments on the options ;)

For the type change, we _could_ have backward compatibility with warning by adding `true` & `false` in the `mkXXXXApply` mapping functions like:
```nix
mkEnumApply {
  always = 21,
  hide = 42,
  true = lib.warn ''something something use "always" now'' 21,
  false = lib.warn ''something something use `"hide"` now'' 42,
}
```

bew commented:
Starting to look good I think!

While looking at the test file I noticed something that might be improved:
```nix
  system.defaults.controlcenter.Sound = "always";
  system.defaults.controlcenter.StageManager = true;
  system.defaults.controlcenter.UserSwitcher = "menuBar";
  system.defaults.controlcenter.WiFi = true;
```
Those lines show how inconsistent the values are for seemingly similar options: some are `true`, others are `"always"` or `"menuBar"`..
👉 Could we avoid having multiple values for the same thing? 🤔

Iamanaws commented:
> Thanks for the changes, looks nice!
> 
> While looking at the test file I noticed something that might be improved:
> 
> ```nix
>   system.defaults.controlcenter.Sound = "always";
>   system.defaults.controlcenter.StageManager = true;
>   system.defaults.controlcenter.UserSwitcher = "menuBar";
>   system.defaults.controlcenter.WiFi = true;
> ```
> 

I updated the test file to use the same values for options that share the same type. 
So now we only apply "both", "always", and true.

Is that what you meant?


> Those lines show how inconsistent the values are for seemingly similar options: some are `true`, others are `"always"` or `"menuBar"`.. 👉 Could we avoid having multiple values for the same thing? 🤔

I leave as bool the ones that are more simple, but I could change them to reflect better the options. I grouped similar stuff and leave just 3 different patterns.

iamanaws committed:
TextInputMenuAgent: add Text Input menu bar option

iamanaws committed:
Spotlight: add menu bar visibility option

iamanaws committed:
controlcenter: add more menu bar defaults

- Use string enums for Control Center menu state options
- Add defaults for AccessibilityShortcuts, AirplayReceiverEnabled,
  Battery, BatteryShowEnergyMode, Hearing, KeyboardBrightness,
  MusicRecognition, ScreenMirroring, StageManager, TimeMachine,
  UserSwitcher, Weather, and WiFi

iamanaws commented:
@Samasaur1 just a friendly ping to bring this up
